### PR TITLE
Fix gRPC Server root span reference issue

### PR DESCRIPTION
**Pull request by kjin**
Calling gRPC server functions would clobber references to root span objects (causing them to either never end, or end at the wrong time) because they are stored in the scope of the patching function (`wrapUnary`, `wrapClientStream`, etc.).

This change fixes that by putting a reference to the root span in the scope of the `func` handler, so a root span can exist per gRPC call.

In addition, unit tests were added/modified:
- to ensure that interleaved gRPC call traces don't interfere with one another.
- to ensure that context is being propagated correctly to call handlers, by creating child spans in these handlers.
- to ensure that gRPC server call traces include the correct subset of stack frames.

### CODE REVIEW

**Comment by kjin** [comment]
@GoogleCloudPlatform/node-team PTAL

File: src/hooks/userspace/hook-grpc.js [hunk 1]
```diff
@@ -160,58 +160,38 @@ function sendMetadataWrapper(rootContext) {
...
         if (agent.config_.enhancedDatabaseReporting) {
-          if (err) {
-            rootContext.addLabel('error', err);
-          } else {
-            rootContext.addLabel('result', JSON.stringify(result));
-          }
-          if (trailer) {
-            rootContext.addLabel('trailing_metadata', JSON.stringify(trailer.getMap()));
-          }
+          shimmer.wrap(call, 'sendMetadata', sendMetadataWrapper(rootContext));
```
Comment (matthewloring) [inline src/hooks/userspace/hook-grpc.js]: Is `call` a unique object per request or is it shared?

**Comment by matthewloring** [comment]
This looks good on my end but it would be good to get one more pair of eyes at least on the test logic due to the complexity.

File: src/hooks/userspace/hook-grpc.js (hunk 1 above)
Comment (kjin) [inline src/hooks/userspace/hook-grpc.js]: It's per request; specifically, it's passed into the request handler and contains information about the request (including the incoming request data, namely `call.request`).

File: src/hooks/userspace/hook-grpc.js [hunk 2]
```diff
@@ -160,58 +160,38 @@ function sendMetadataWrapper(rootContext) {
...
-  });
-  shimmer.wrap(handlerSet, 'serialize', function (serialize) {
-    return function serializeTrace() {
-      var serializeResult = serialize.apply(this, arguments);
-      endRootSpanForRequest(rootContext);
-      return serializeResult;
+        // args[1] is the callback.
+        // Here, we patch the callback so that the span is ended immediately
+        // beforehand.
+        args[1] = function (err, result, trailer, flags) {
```
Comment (cristiancavalli) [inline src/hooks/userspace/hook-grpc.js]: Disclaimer: personal opinion
Overwriting arguments make me nervous one could achieve the same result:

```
var applyArgs = [].concat(args);
applyArgs[1]  = function () {...}
```

File: src/hooks/userspace/hook-grpc.js [hunk 3]
```diff
@@ -224,46 +204,48 @@ function wrapUnary(namespace, handlerSet, requestName) {
...
+        // event handlers.
+        stream.on('finish', function () {
+          // End the span unless there is an error. (If there is, the span will
+          // be ended in the error event handler. This is to ensure that the
+          // 'error' label is applied.)
+          if (stream.status.code === 0) {
+            endSpan();
+          }
+        });
+        stream.on('error', function (err) {
```
Comment (cristiancavalli) [inline src/hooks/userspace/hook-grpc.js]: `on` actually returns the instance for chaining so one could do:

`eventEmitter.on(...).on(...);`

File: src/hooks/userspace/hook-grpc.js (hunk 2 above)
Comment (kjin) [inline src/hooks/userspace/hook-grpc.js]: Sounds good, will do

File: src/hooks/userspace/hook-grpc.js (hunk 3 above)
Comment (kjin) [inline src/hooks/userspace/hook-grpc.js]: Did not know this - but I think I think I would prefer the slight clarity of two `stream.on`s to the slight succinctness of having one.

**Pull request merged**